ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java

```diff
@@ -0,0 +1,221 @@
+        log.debug("Reordered tensor outputs size is {}", sortedMap.size());
+        if (tensorOutputs.size() == 1) {
+            // batch API case
+            int status = tensorOutputs.get(0).getStatusCode();
+            if (status == HttpStatus.SC_OK) {
+                modelTensors.add(tensorOutputs.get(0));
+                actionListener.onResponse(modelTensors);
+            } else {
+                try {
+                    actionListener
+                        .onFailure(
+                            new OpenSearchStatusException(
+                                AccessController
+                                    .doPrivileged(
+                                        (PrivilegedExceptionAction<String>) () -> GSON
+                                            .toJson(tensorOutputs.get(0).getMlModelTensors().get(0).getDataAsMap())
+                                    ),
+                                RestStatus.fromCode(status)
+                            )
+                        );
+                } catch (PrivilegedActionException e) {
+                    actionListener.onFailure(new OpenSearchStatusException(e.getMessage(), RestStatus.fromCode(statusCode)));
+                }
+            }
+        } else {
+            // non batch API.
+            for (Map.Entry<Integer, ModelTensors> entry : sortedMap.entrySet()) {
+                modelTensors.add(entry.getKey(), entry.getValue());
+            }
+            actionListener.onResponse(modelTensors);
+        }
+    }
+
+    protected class MLResponseSubscriber implements Subscriber<ByteBuffer> {
+        private Subscription subscription;
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            this.subscription = s;
+            s.request(Long.MAX_VALUE);
+        }
+
+        @Override
+        public void onNext(ByteBuffer byteBuffer) {
+            responseBody.append(StandardCharsets.UTF_8.decode(byteBuffer));
+            subscription.request(Long.MAX_VALUE);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            log
+                .error(
+                    "Error on receiving response body from remote: {}",
+                    t instanceof NullPointerException ? "NullPointerException" : t.getMessage(),
+                    t
+                );
+            processResponse(statusCode, responseBody.toString(), parameters, tensorOutputs);
+            countDownLatch.getCountDownLatch().countDown();
+            response(tensorOutputs);
+        }
+
+        @Override
+        public void onComplete() {
+            processResponse(statusCode, responseBody.toString(), parameters, tensorOutputs);
+            countDownLatch.getCountDownLatch().countDown();
+            response(tensorOutputs);
+        }
+    }
+
+    private void response(Map<Integer, ModelTensors> tensors) {
+        if (countDownLatch.getCountDownLatch().getCount() == 0) {
```

Comment:
Does this mean we are done with getting all the embeddings? If yes, can we add a comment?

Reply:
sure